common/solver/batch_gmres_kernels.hpp.inc

```diff
@@ -0,0 +1,870 @@
+                    apply_plane_rotation(
+                        &cs_shared_entry.values[k * cs_shared_entry.stride],
+                        &sn_shared_entry.values[k * sn_shared_entry.stride],
+                        nrhs,
+                        &H_shared_entry.values[k * H_shared_entry.stride +
+                                               inner_iter * nrhs],
+                        &H_shared_entry.values[(k + 1) * H_shared_entry.stride +
+                                               inner_iter * nrhs],
+                        converged);
+                    __syncthreads();
+                }
+
+
+                // compute sine and cos
+                generate_plane_rotation(
+                    &H_shared_entry.values[inner_iter * H_shared_entry.stride +
+                                           inner_iter * nrhs],
+                    &H_shared_entry
+                         .values[(inner_iter + 1) * H_shared_entry.stride +
+                                 inner_iter * nrhs],
+                    nrhs,
+                    &cs_shared_entry
+                         .values[inner_iter * cs_shared_entry.stride],
+                    &sn_shared_entry
+                         .values[inner_iter * cs_shared_entry.stride],
+                    converged);
+                __syncthreads();
+
+
+                // temp = cs(inner_iter) * s(inner_iter)
+                // s(inner_iter + 1) = -1 * conj(sn(inner_iter)) * s(inner_iter)
+                // s(inner_iter) = temp
+                // Hessenberg(inner_iter , inner_iter) = cs(inner_iter) *
+                // Hessenberg(inner_iter , inner_iter) + sn(inner_iter) *
+                // Hessenberg(inner_iter + 1, inner_iter)
+                // Hessenberg(inner_iter + 1, inner_iter) = 0
+                apply_plane_rotation(
+
+                    &cs_shared_entry
+                         .values[inner_iter * cs_shared_entry.stride],
+                    &sn_shared_entry
+                         .values[inner_iter * sn_shared_entry.stride],
+                    nrhs,
+                    &s_shared_entry.values[inner_iter * s_shared_entry.stride],
+                    &s_shared_entry
+                         .values[(inner_iter + 1) * s_shared_entry.stride],
+                    converged);
+
+                apply_plane_rotation(
+
+                    &cs_shared_entry
+                         .values[inner_iter * cs_shared_entry.stride],
+                    &sn_shared_entry
+                         .values[inner_iter * sn_shared_entry.stride],
+                    nrhs,
+                    &H_shared_entry.values[inner_iter * H_shared_entry.stride +
+                                           inner_iter * nrhs],
+                    &H_shared_entry
+                         .values[(inner_iter + 1) * H_shared_entry.stride +
+                                 inner_iter * nrhs],
+                    converged);
```

Comment:
For readability, this makes sense. For performance, though it might not matter too much here, you can do both of these in one loop.